### ISSUE TITLE
Running integration tests with subproject dependency yields NoClassDefFound

### ISSUE THREAD

**Issue by joemarner** [eclipse/buildship#710]
Hi,

I have two subprojects a and b where a depends on classes of b. When running a test in project a that depends on classes of project b in eclipse via "Run as JUnit Test" , I get a NoClassDefFound b exception. Running the test via gradlew on the command line works fine. Using the gradle task in eclipse ("Run as gradle Test") seems to not execute any test at all.

Setup:
jdk-9.0.4
eclipse 4.7.3a
Gradle 4.6, also tried with 4.8.1
Buildship: 2.2.1v20180125-1441

sample project attached (test.zip).

Thanks for any help!

[test.zip](https://github.com/eclipse/buildship/files/2132582/test.zip)

**Comment by oehme** [comment]
You used `compile project(':b')`, but didn't specify that `intTestCompile` should extend `testCompile`. Either use `implementation project(':b')` or add the missing extends clause.

**Comment by oehme** [comment]
@donat I noticed that we don't create classpath attributes for project dependencies. Does Buildship wrongly filter them even though they don't have these attributes? Also, shouldn't they have these attributes?

**Comment by joemarner** [comment]
Hi,

hanks for your fast reply. I tried your hint with adding testCompile project(':b') to the dependencies, but this did not solve it for me.

Also https://docs.gradle.org/current/userguide/java_plugin.html says, that testCompile inherits from compile (and that both are deprecated and are replaced by implementation). So a dependency to b in compile should also be valid for testCompile.

Strange enough, the runtime configuration of the integration test shows a project dependency of project b (wanted to attach screenshot, but I keep having problems with attaching stuff in github).

And: Compiling is not an issue, just running it.

**Comment by oehme** [comment]
Nevermind my earlier comment, it wasn't the problem :)

**Comment by donat** [comment]
The missing resource is due to the classpath separation we introduced in [Buildship 2.2](https://projects.eclipse.org/projects/tools.buildship/releases/2.2.0). Unfortunately, it's only an approximation to what Gradle does, due to the limitation of Eclipse JDT.

The separation in Eclipse is done based on the source sets. If you run a test from the `test` source set, then all classes will be loaded from the `test` and the _used_ source sets (`main` in this case). The source set is part of the _used_ source sets if the runtime classpath [contains the output folder](https://github.com/gradle/gradle/blob/c052c1865d8cea4ac0fc8a4e9be2da8d8d9da297/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java#L195). Since your `intTest` is only available on project `a`, the used source set calculation fails. 

@joemarner  sorry about the technical mumbo jumbo, this description is more for us for future reference. Just move the `intTest` declaration into `subprojects {}` and refresh your project.

**Comment by oehme** [comment]
@donat We should probably improve that by detecting that `intTest` is an extension of `main` and should thus see everything that `main` can see. I think that would have fixed this particular issue.

**Comment by donat** [comment]
Agreed. We could probably do it together with https://github.com/eclipse/buildship/issues/689.

**Comment by robross0606** [comment]
@donat I'm facing a similar problem and trying to work around the issue.  Your suggestion to @joemarner a couple posts back focuses on `intTest` but I believe the same problem happens for regular `test` sources in the standard `src\test\java\` directory of project **a**.  What is the suggested workaround for situations where you are not declaring a custom source set and are using standard configuration by convention?
